File: leetcode/Wildcard Matching.cpp
class Solution {
public:
	bool isMatch(const char *s, const char *p) {
		bool star = false;
		const char *str = NULL, *ptr = NULL;
		for (str = s, ptr = p; *str; ++ str, ++ ptr) {
			switch (*ptr) {
			case '?':
				break;
			case '*':
				star = true;
				s = str;
				p = ptr;
				while ('*' == *p) {
					++ p;
				}
				if ('\0' == *p) {
					return true;
				}
				str = s - 1;
				ptr = p - 1;
				break;
			default:
				if (*str != *ptr) {
					if (!star) {
						return false;
					}
					++ s;
					str = s - 1;
					ptr = p - 1;
				}
			}
		}
		while ('*' == *ptr) {
			++ ptr;
		}
		return '\0' == *ptr;
	}
};

File: answers/leetcode/Wildcard Matching/Wildcard Matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] dp = new boolean[s.length() + 1][p.length() + 1];
        dp[0][0] = true;
        for (int i = 0; i <= s.length(); i++) {
            for (int j = 1; j <= p.length(); j++) {
                switch (p.charAt(j - 1)) {
                    case '?':
                        dp[i][j] = i > 0 ? dp[i - 1][j - 1] : false;
                        break;
                    case '*':
                        dp[i][j] = dp[i][j - 1] || (i > 0 ? dp[i - 1][j] : false);
                        break;
                    default:
                        dp[i][j] = i > 0 ? dp[i - 1][j - 1] && s.charAt(i - 1) == p.charAt(j - 1) : false;
                }
            }
        }
        return dp[s.length()][p.length()];
    }
}